File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointList/MountPointsList.java
package ru.nsu.bobrofon.easysshfs.mountpointList;


import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.LinkedList;
import java.util.List;

import ru.nsu.bobrofon.easysshfs.log.LogSingleton;
import ru.nsu.bobrofon.easysshfs.mountpointList.mountpoint.MountPoint;

public class MountPointsList {
	private final static String TAG = "MOUNT_POINTS_LIST";
	private final static String STORAGE_FILE = "mountpoints";

	private List<MountPoint> mMountPoints;

	private static MountPointsList self;

	private MountPointsList() {
	}

	public List<MountPoint> getMountPoints() {
		return mMountPoints;
	}

	public void autoMount(final Context context) {
		for (final MountPoint item : mMountPoints) {
			if (item.getAutoMount()/* && !item.isMounted()*/) {
				item.mount(context);
			}
		}
	}

	public void umount(final Context context) {
		for (final MountPoint item: mMountPoints) {
			item.umount(false, context);
		}
	}

	public void registerObserver(final MountPoint.Observer observer, final Context context) {
		for (final MountPoint item : mMountPoints) {
			item.registerObserver(observer, context);
		}
	}

	public void unregisterObserver(final MountPoint.Observer observer) {
		for (final MountPoint item : mMountPoints) {
			item.unregisterObserver(observer);
		}
	}

	private MountPointsList load(final Context context) {
		final SharedPreferences settings = context.getSharedPreferences(STORAGE_FILE, 0);
		try {
			JSONArray selfJson = new JSONArray(settings.getString(STORAGE_FILE, "[]"));

			mMountPoints = new LinkedList<>();
			for (int i = 0; i < selfJson.length(); ++i) {
				MountPoint mountPoint = new MountPoint();
				mountPoint.json(selfJson.getJSONObject(i));
				mMountPoints.add(mountPoint);
			}
		} catch (JSONException e) {
			Log.e(TAG, e.getMessage());
			log(e.getMessage());
		}

		return this;
	}

	public void save(final Context context) {
		final SharedPreferences settings = context.getSharedPreferences(STORAGE_FILE, 0);

		JSONArray selJson = new JSONArray();
		for (MountPoint item : mMountPoints) {
			selJson.put(item.json());
		}

		SharedPreferences.Editor prefsEditor = settings.edit();
		prefsEditor.putString(STORAGE_FILE, selJson.toString()).apply();
	}

	private void log(final CharSequence message) {
		LogSingleton.getLogModel().addMessage(message);
	}

	public static MountPointsList getIntent(final Context context) {
		if (self == null) {
			self = new MountPointsList().load(context);
		}
		return self;
	}

}

File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointlist/MountPointsList.kt
package ru.nsu.bobrofon.easysshfs.mountpointlist

import java.lang.ref.WeakReference
import java.util.LinkedList

import android.content.Context
import android.util.Log

import org.json.JSONArray
import org.json.JSONException

import com.topjohnwu.superuser.Shell

import ru.nsu.bobrofon.easysshfs.log.AppLog
import ru.nsu.bobrofon.easysshfs.mountpointlist.mountpoint.MountPoint
import ru.nsu.bobrofon.easysshfs.mountpointlist.mountpoint.MountStateChangeObserver

class MountPointsList(
    mountPoints: MutableList<MountPoint> = LinkedList(),
    private val appLog: AppLog = AppLog.instance()
) {
    var mountPoints: MutableList<MountPoint> = mountPoints
        private set

    private val autoMountObservable = AutoMountObservable()

    fun needAutomount(): Boolean = mountPoints.any {
        it.autoMount && !it.isMounted
    }

    fun autoMount(shell: Shell) = mountPoints.forEach {
        if (it.autoMount && !it.isMounted) {
            it.mount(shell)
        }
    }

    val isAutoMountEnabled get() = mountPoints.any { it.autoMount }

    fun umount(shell: Shell) = mountPoints.forEach { it.umount(shell) }

    fun registerMountObserver(observer: MountStateChangeObserver) = mountPoints.forEach {
        it.registerObserver(observer)
    }

    fun unregisterMountObserver(observer: MountStateChangeObserver) = mountPoints.forEach {
        it.unregisterObserver(observer)
    }

    fun registerAutoMountObserver(observer: AutoMountChangeObserver) {
        autoMountObservable.registerObserver(observer, isAutoMountEnabled)
    }

    fun unregisterAutoMountObserver(observer: AutoMountChangeObserver) {
        autoMountObservable.unregisterObserver(observer)
    }

    private fun load(context: Context) {
        val settings = context.getSharedPreferences(STORAGE_FILE, 0)
        try {
            val selfJson = JSONArray(settings.getString(STORAGE_FILE, "[]"))

            mountPoints.clear()
            for (i in 0 until selfJson.length()) {
                val mountPoint = MountPoint()
                mountPoint.json(selfJson.getJSONObject(i))
                mountPoints.add(mountPoint)
            }
        } catch (e: JSONException) {
            Log.e(TAG, e.message ?: "")
            log(e.message.orEmpty())
        }
    }

    fun save(context: Context) {
        val settings = context.getSharedPreferences(STORAGE_FILE, 0)

        val selJson = JSONArray()
        for (item in mountPoints) {
            selJson.put(item.json())
        }

        val prefsEditor = settings.edit()
        prefsEditor.putString(STORAGE_FILE, selJson.toString()).apply()
        autoMountObservable.notifyChanged(isAutoMountEnabled)
    }

    private fun log(message: CharSequence) {
        appLog.addMessage(message)
    }

    companion object {
        private const val TAG = "MOUNT_POINTS_LIST"
        private const val STORAGE_FILE = "mountpoints"

        private var instance = WeakReference<MountPointsList?>(null)

        @Synchronized
        fun instance(context: Context): MountPointsList {
            val oldInstance = instance.get()
            if (oldInstance == null) {
                val newInstance = MountPointsList().apply { load(context) }
                instance = WeakReference(newInstance)
                return newInstance
            }
            return oldInstance
        }
    }

}
